### PR TITLE
Normalize patterns before hash consing them for analysis

### LINKED ISSUE #12489: Exponential compile times when pattern matching on type alias
## reproduction steps

using Scala 2.13.5,

```scala
// Note: Removing `+` here has no effect on the issue:
type ~[+A, +B] = (A, B)

// Dummy type to allow us to map + pattern match
trait A[T] { def map[U](fn: T => U): A[U] = new A[U] {} }

// 25 nested Int's, both the same type
type B = ((((((((((((((((((((((((Int, Int), Int), Int), Int), Int), Int), Int), Int), Int), Int), Int), Int), Int), Int), Int), Int), Int), Int), Int), Int), Int), Int), Int), Int)
type C = Int ~ Int ~Int ~Int ~Int ~Int ~Int ~Int ~Int ~Int ~Int ~Int ~Int ~Int ~Int ~Int ~Int ~Int ~Int ~Int ~Int ~Int ~Int ~Int~Int

val ab = new A[B] {}
val ac = new A[C] {}

// Instant compile time:
ab.map { case ((((((((((((((((((((((((_, _), _), _), _), _), _), _), _), _), _), _), _), _), _), _), _), _), _), _), _), _), _), _), _) => 7 }

// Exponential compile time:
ac.map { case ((((((((((((((((((((((((_, _), _), _), _), _), _), _), _), _), _), _), _), _), _), _), _), _), _), _), _), _), _), _), _) => 7 }
```

## problem

Deconstructing a simple alias for a type in a case match causes exponentially increasing compile times, dependent on the depth of the deconstruction.

In the code above, `type B` and `type C` are the same, but `C` is constructed through aliases. Deconstructing type `C` takes vastly longer than deconstructing type `B`.

I would expect that `ab.map...` and `ac.map...` should take the same length of time to compile.

### PR DESCRIPTION
**Pull request by joroKr21**
It increases the chance of structural equality and reduces the cost of type equality comparison.
This should fix scala/bug#12489 but I'm not sure if we can add a test for it.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/transform/patmat/Logic.scala [hunk 1]
```diff
@@ -752,17 +752,15 @@ trait ScalaLogic extends Interface with Logic with TreeAndTypeAnalysis {
...
-              debug.patmat("unique const: "+ ((tp, c)))
-              c
-            case _ =>
-              val fresh = mkFresh
-              debug.patmat("uniqued const: "+ ((tp, fresh)))
-              uniques(tp) = fresh
-              fresh
+        uniques.getOrElse(tp, {
+          // normalize to increase the chance of structural equality and reduce the cost of =:=
+          val normalized = tp.map(_.normalize)
```
Comment (dwijnand) [inline src/compiler/scala/tools/nsc/transform/patmat/Logic.scala:757]: Why not tp.normalize?

**Review by dwijnand — APPROVED** [review]
I think this is good, just a question.

File: src/compiler/scala/tools/nsc/transform/patmat/Logic.scala (hunk 1 above)
Comment (joroKr21) [reply src/compiler/scala/tools/nsc/transform/patmat/Logic.scala:757]: `normalize` is shallow - it will only dealias the top-level type alias right?

**Comment by SethTisue** [comment]
@dwijnand is the answer to your question satisfactory? (want to go ahead and hit "merge"?)

File: src/compiler/scala/tools/nsc/transform/patmat/Logic.scala (hunk 1 above)
Comment (joroKr21) [reply src/compiler/scala/tools/nsc/transform/patmat/Logic.scala:757]: Yes indeed:
```scala
scala> type A = Int
type A

scala> type F[x] = List[x]
type F

scala> typeOf[F[A]].normalize
val res0: $r.intp.global.Type = List[A]
```

**Pull request merged**